apps/andi/lib/andi/input_schemas/ingestions/extract_steps/extract_http_step.ex

```diff
@@ -22,59 +22,72 @@ defmodule Andi.InputSchemas.Ingestions.ExtractHttpStep do
-    |> validate_url()
-    |> validate_required(@required_fields, message: "is required")
-    |> validate_key_value_set(:headers)
-    |> validate_key_value_set(:queryParams)
+      |> Changeset.cast(changes_with_id, @cast_fields, empty_values: [], force_changes: true)
+      |> Changeset.cast_embed(:headers, with: &ExtractHeader.changeset/2)
+      |> Changeset.cast_embed(:queryParams, with: &ExtractQueryParam.changeset/2)
   end
 
-  def changeset_for_draft(extract_step, changes) do
-    changes_with_id = StructTools.ensure_id(extract_step, changes)
-
-    extract_step
-    |> cast(changes_with_id, @cast_fields, empty_values: [])
-    |> cast_embed(:headers, with: &ExtractHeader.changeset_for_draft/2)
-    |> cast_embed(:queryParams, with: &ExtractQueryParam.changeset_for_draft/2)
+  def validate(extract_step_changeset) do
+    data_as_changes =
+      extract_step_changeset
+      |> Changeset.apply_changes()
+      |> StructTools.to_map()
+      |> format()
+
+    validated_extract_step_changeset = extract_step_changeset
+      |> Map.replace(:errors, [])
+      |> Changeset.cast(data_as_changes, @cast_fields, empty_values: [], force_changes: true)
+      |> Changeset.cast_embed(:headers, with: &ExtractHeader.changeset/2)
+      |> Changeset.cast_embed(:queryParams, with: &ExtractQueryParam.changeset/2)
+      |> validate_body_format()
+      |> validate_url()
+      |> Changeset.validate_required(@required_fields, message: "is required")
+      |> validate_key_value_set(:headers)
+      |> validate_key_value_set(:queryParams)
+
+    if is_nil(Map.get(validated_extract_step_changeset, :action, nil)) do
+      Map.put(validated_extract_step_changeset, :action, :display_errors)
+    else
+      validated_extract_step_changeset
+    end
   end
 
-  def changeset_from_form_data(form_data) do
-    form_data_as_params =
-      form_data
-      |> AtomicMap.convert(safe: false, underscore: false)
-      |> Map.put_new(:queryParams, %{})
-      |> Map.put_new(:headers, %{})
-      |> convert_map_with_index_to_list(:queryParams)
-      |> convert_map_with_index_to_list(:headers)
+  def preload(struct), do: StructTools.preload(struct, [:headers, :queryParams])
 
-    changeset(form_data_as_params)
+  defp format(changes) do
+    changes
+      |> format_url()
   end
 
-  def changeset_from_andi_step(nil), do: changeset(%{})
+  defp format_url(%{url: url, queryParams: %{} = query_params} = changes) do
+    query_param_array = Enum.reduce(Map.to_list(query_params), [], fn {_index, query_param}, acc ->
+      acc ++ [%{key: query_param.key, value: query_param.value}]
+    end)
+    new_url = Andi.URI.update_url_with_params(url, query_param_array)
+    Map.put(changes, :url, new_url)
+  end
 
-  def changeset_from_andi_step(dataset_extract_step) do
-    dataset_extract_step
-    |> StructTools.to_map()
-    |> AtomicMap.convert(safe: false, underscore: false)
-    |> changeset()
+  defp format_url(%{url: url, queryParams: query_params} = changes) do
```

Comment:
all this does is tack on the query params to the url, which is to be shown via the ui, since the ui's url field gets fed by this url, I think it is fine on both ends. I can try to move the concat to a UI level, but I think that might get messy? (although maybe not by much, since auth step already has something)

Reply:
I just tested locally by calling it twice and it seems to handle it properly. No issues here!